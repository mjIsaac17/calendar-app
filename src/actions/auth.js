import { fetchNoToken, fetchToken } from "../helpers/fetch";
import { types } from "../types/types";

const baseEndpoint = "auth";

export const startLogin = (email, password) => {
  //the return is triggered by thunk when we need to use async functions
  return async (dispatch) => {
    const resp = await fetchNoToken(baseEndpoint, { email, password }, "POST");
    const body = await resp.json();
    if (resp.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());
      dispatch(login({ uid: body.uid, name: body.name }));
      return "";
    }
    // else{
    //     //Using sweet alert
    //     //Swal.fire
    // }
    return body.msg;
  };
};

export const startRegister = (email, password, name) => {
  return async (dispatch) => {
    const resp = await fetchNoToken(
      `${baseEndpoint}/new`,
      { email, password, name },
      "POST"
    );
    const body = await resp.json();
    if (resp.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());
      dispatch(login({ uid: body.uid, name: body.name }));
      return "";
    }
    return body.msg;
  };
};

//Renew token
export const startChecking = () => {
  return async (dispatch) => {
    const resp = await fetchToken(`${baseEndpoint}/renew`);
    const body = await resp.json();
    if (resp.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());
      dispatch(login({ uid: body.uid, name: body.name }));
    } else dispatch(checkingFinish());
  };
};

const checkingFinish = () => ({ type: types.authCheckingFinish });

const login = (user) => ({
  type: types.authLogin,
  payload: user,
});
