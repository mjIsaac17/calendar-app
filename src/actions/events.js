import { fetchToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types";

export const eventStartAddNew = (event) => {
  return async (dispatch, getState) => {
    try {
      const { uid, name } = getState().auth;
      const resp = await fetchToken("events", event, "POST");
      const body = await resp.json();
      if (resp.ok) {
        event.id = body.event.id;
        event.user = {
          _id: uid,
          name,
        };
        dispatch(eventAddNew(event));
      }
    } catch (error) {
      console.log(error);
    }
  };
};

const eventAddNew = (event) => ({
  type: types.eventAddNew,
  payload: event,
});

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event,
});

export const eventClearActive = () => ({
  type: types.eventClearActive,
});

export const eventStartUpdate = (event) => {
  return async (dispatch) => {
    try {
      const resp = await fetchToken(`events/${event.id}`, event, "PUT");
      const body = await resp.json();
      if (resp.ok) {
        dispatch(eventUpdate(event));
      } else {
        alert(`Error. ${body.msg}`);
      }
    } catch (error) {
      console.log(error);
    }
  };
};

const eventUpdate = (event) => ({
  type: types.eventUpdate,
  payload: event,
});

export const eventStartDelete = () => {
  return async (dispatch, getState) => {
    try {
      const { id } = getState().calendar.activeEvent;
      const resp = await fetchToken(`events/${id}`, {}, "DELETE");
      const body = await resp.json();
      if (resp.ok) {
        dispatch(eventDelete());
      } else {
        alert(`Error. ${body.msg}`);
      }
    } catch (error) {
      console.log(error);
    }
  };
};

const eventDelete = () => ({
  type: types.eventDelete,
});

export const eventStartLoading = () => {
  return async (dispatch) => {
    try {
      const resp = await fetchToken("events");
      const body = await resp.json();
      const events = prepareEvents(body.events);
      dispatch(eventLoaded(events));
    } catch (error) {
      console.log(error);
    }
  };
};

const eventLoaded = (events) => ({
  type: types.eventLoaded,
  payload: events,
});

export const eventLogout = () => ({
  type: types.eventLogout,
});
