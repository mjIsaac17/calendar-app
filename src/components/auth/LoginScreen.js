import React from "react";
import { useState } from "react";
import { useDispatch } from "react-redux";
import { startLogin, startRegister } from "../../actions/auth";
import { useForm } from "../../hooks/useForm";
import "./login.css";

export const LoginScreen = () => {
  const dispatch = useDispatch();

  const [errorMessageLogin, setErrorMessageLogin] = useState("");
  const [errorMessageRegister, setErrorMessageRegister] = useState("");

  const [formLoginValues, handleInputChange] = useForm({
    lEmail: "isaac@gmail.com", //loginEmail
    lPassword: "123456",
  });
  const { lEmail, lPassword } = formLoginValues;

  const [formRegisterValues, handleRegisterInputChange] = useForm({
    rName: "Caasi",
    rEmail: "caasi@gmail.com", //registerEmail
    rPassword1: "123456",
    rPassword2: "123456",
  });

  const { rName, rEmail, rPassword1, rPassword2 } = formRegisterValues;

  const handleLogin = async (e) => {
    e.preventDefault();
    const errorMessage = await dispatch(startLogin(lEmail, lPassword));
    setErrorMessageLogin(errorMessage);
  };

  const handleRegister = async (e) => {
    e.preventDefault();
    if (rPassword1 !== rPassword2) {
      return setErrorMessageRegister("Passwords must be the same");
    }
    //More validations if required
    const errorMessage = await dispatch(
      startRegister(rEmail, rPassword1, rName)
    );
    setErrorMessageRegister(errorMessage);
  };

  return (
    <div className="container login-container">
      <div className="row">
        <div className="col-md-6 login-form-1">
          <h3>Login</h3>
          <form onSubmit={handleLogin}>
            {errorMessageLogin !== "" && (
              <div className="alert alert-danger">
                <p className="noMargin text-center">{errorMessageLogin}</p>
              </div>
            )}
            <div className="form-group">
              <input
                name="lEmail"
                type="text"
                className="form-control"
                placeholder="Email"
                value={lEmail}
                onChange={handleInputChange}
              />
            </div>
            <div className="form-group">
              <input
                name="lPassword"
                value={lPassword}
                type="password"
                className="form-control"
                placeholder="Password"
                onChange={handleInputChange}
              />
            </div>
            <div className="form-group">
              <input type="submit" className="btnSubmit" value="Login" />
            </div>
          </form>
        </div>

        <div className="col-md-6 login-form-2">
          <h3>Register</h3>
          <form onSubmit={handleRegister}>
            {errorMessageRegister !== "" && (
              <div className="alert alert-danger">
                <p className="noMargin text-center">{errorMessageRegister}</p>
              </div>
            )}
            <div className="form-group">
              <input
                name="rName"
                value={rName}
                onChange={handleRegisterInputChange}
                type="text"
                className="form-control"
                placeholder="Nombre"
              />
            </div>
            <div className="form-group">
              <input
                name="rEmail"
                value={rEmail}
                onChange={handleRegisterInputChange}
                type="email"
                className="form-control"
                placeholder="Correo"
              />
            </div>
            <div className="form-group">
              <input
                name="rPassword1"
                value={rPassword1}
                onChange={handleRegisterInputChange}
                type="password"
                className="form-control"
                placeholder="Contraseña"
              />
            </div>

            <div className="form-group">
              <input
                name="rPassword2 "
                value={rPassword2}
                onChange={handleRegisterInputChange}
                type="password"
                className="form-control"
                placeholder="Repita la contraseña"
              />
            </div>

            <div className="form-group">
              <input type="submit" className="btnSubmit" value="Crear cuenta" />
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};
