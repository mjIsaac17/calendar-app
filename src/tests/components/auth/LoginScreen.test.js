import thunk from "redux-thunk";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { LoginScreen } from "../../../components/auth/LoginScreen";
import { startLogin, startRegister } from "../../../actions/auth";
import { act, waitFor } from "@testing-library/react";

jest.mock("../../../actions/auth", () => ({
  startLogin: jest.fn(),
  startRegister: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {};
const store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <LoginScreen />
  </Provider>
);

describe("Tests in <LoginScreen />", () => {
  test("should load the component correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should call the login dispatch, startLogin", async () => {
    wrapper.find('input[name="lEmail"]').simulate("change", {
      target: { name: "lEmail", value: "test@gmail.com" },
    });
    wrapper.find('input[name="lPassword"]').simulate("change", {
      target: { name: "lPassword", value: "123456" },
    });

    wrapper.find("form").at(0).prop("onSubmit")({ preventDefault() {} });
    // waitFor is added because startlogin returns an errorMessage, and it is an async fn
    await waitFor(() => {
      expect(startLogin).toHaveBeenCalledWith("test@gmail.com", "123456");
    });
  });

  test("should not call the register an user the passwords are different", () => {
    wrapper.find('input[name="rName"]').simulate("change", {
      target: { name: "rName", value: "testName" },
    });
    wrapper.find('input[name="rEmail"]').simulate("change", {
      target: { name: "rEmail", value: "test@gmail.com" },
    });
    wrapper.find('input[name="rPassword1"]').simulate("change", {
      target: { name: "rPassword1", value: "123456" },
    });
    wrapper.find('input[name="rPassword2"]').simulate("change", {
      target: { name: "rPassword2", value: "654321" },
    });
    act(() => {
      wrapper.find("form").at(1).prop("onSubmit")({ preventDefault() {} });
    });

    expect(startRegister).toHaveBeenCalledTimes(0);
  });

  test("should call the register with same passwords", async () => {
    wrapper.find('input[name="rPassword2"]').simulate("change", {
      target: { name: "rPassword2", value: "123456" },
    });
    await waitFor(() => {
      wrapper.find("form").at(1).prop("onSubmit")({ preventDefault() {} });
    });
    expect(startRegister).toHaveBeenCalledTimes(1);
  });
});
