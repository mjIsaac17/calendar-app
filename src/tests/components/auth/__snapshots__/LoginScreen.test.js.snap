// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <LoginScreen /> should load the component correctly 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Login
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lEmail"
            onChange={[Function]}
            placeholder="Email"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lPassword"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Register
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rName"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword1"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
